src/webgpu/shader/execution/builtin/clamp.spec.ts

```diff
@@ -0,0 +1,165 @@
+        // This is enforcing that low <= high, since most backends languages
+        // have undefined behaviour for high < low, so implementations would
+        // need to polyfill, and it is unclear if this was intended.
+        //
+        // https://github.com/gpuweb/gpuweb/issues/2557 discusses changing the
+        // spec to explicitly require low <= high.
+        if (fi <= gi) {
+          // Intentionally not using clamp from math.ts or other TypeScript
+          // defined clamp to avoid rounding issues from going in/out of
+          // `number` type.
+          const precise_expected = ei <= fi ? f : ei <= gi ? e : g;
+          const expected = isSubnormalScalar(precise_expected)
+            ? anyOf(precise_expected, f32(0.0))
+            : precise_expected;
+          cases.push({ input: [e, f, g], expected });
+        }
+      });
+    });
+  });
+  return cases;
+}
+
+g.test('integer_builtin_functions,unsigned_clamp')
+  .uniqueId('386458e12e52645b')
+  .specURL('https://www.w3.org/TR/2021/WD-WGSL-20210929/#integer-builtin-functions')
+  .desc(
+    `
+unsigned clamp:
+T is u32 or vecN<u32> clamp(e: T , low: T, high: T) -> T Returns min(max(e, low), high). Component-wise when T is a vector. (GLSLstd450UClamp)
+
+Please read the following guidelines before contributing:
+https://github.com/gpuweb/cts/blob/main/docs/plan_autogen.md
+`
+  )
+  .params(u =>
+    u
+      .combine('storageClass', ['uniform', 'storage_r', 'storage_rw'] as const)
+      .combine('vectorize', [undefined, 2, 3, 4] as const)
+  )
+  .fn(async t => {
+    const cfg: Config = t.params;
+    cfg.cmpFloats = correctlyRoundedThreshold();
+
+    // This array must be strictly increasing, since that ordering determines
+    // the expected values.
+    const test_values: Array<Scalar> = [
+      u32(0),
+      u32(1),
+      u32(2),
+      u32(0x70000000),
+      u32(0x80000000),
+      u32(0xffffffff),
+    ];
+
+    run(t, 'clamp', [TypeU32, TypeU32, TypeU32], TypeU32, cfg, generateTestCases(test_values));
+  });
+
+g.test('integer_builtin_functions,signed_clamp')
+  .uniqueId('da51d3c8cc902ab2')
+  .specURL('https://www.w3.org/TR/2021/WD-WGSL-20210929/#integer-builtin-functions')
+  .desc(
+    `
+signed clamp:
+T is i32 or vecN<i32> clamp(e: T , low: T, high: T) -> T Returns min(max(e, low), high). Component-wise when T is a vector. (GLSLstd450SClamp)
+
+Please read the following guidelines before contributing:
+https://github.com/gpuweb/cts/blob/main/docs/plan_autogen.md
+`
+  )
+  .params(u =>
+    u
+      .combine('storageClass', ['uniform', 'storage_r', 'storage_rw'] as const)
+      .combine('vectorize', [undefined, 2, 3, 4] as const)
+  )
+  .fn(async t => {
+    const cfg: Config = t.params;
+    cfg.cmpFloats = correctlyRoundedThreshold();
+
+    // This array must be strictly increasing, since that ordering determines
+    // the expected values.
+    const test_values: Array<Scalar> = [
+      i32Bits(0x80000000),
```

Comment:
Any reason you're not using the constants for int-min here?

Reply:
Done